### PR TITLE
(DOCSP-26563): Kotlin Custom Function Authentication

### PR DESCRIPTION
**Pull request by cbullinger**
## Pull Request Info

- Add Custom Function section to existing Authenticate Users page
- Small QC updates to Authenticate Users page

### Jira

- https://jira.mongodb.org/browse/DOCSP-26563

### Staged Changes

- [Authenticate Users - Kotlin SDK](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-26563-function-creds/sdk/kotlin/app-services/authenticate-users/)

### Reminder Checklist

If your PR modifies the docs, you might need to also update some corresponding
pages. Check if completed or N/A.

- [x] Create Jira ticket for corresponding docs-app-services update(s), if any: https://jira.mongodb.org/browse/DOCSP-28243
- [x] Checked/updated Admin API
- [x] Checked/updated CLI reference

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

**Comment by clementetb** [comment]
Looking good. Should we also document the new User metadata functions (see `UserExt.kt`): 

`User.profileAsBsonDocument` and `User.customDataAsBsonDocument`.

**Review by clementetb — APPROVED** [review]

**Comment by cbullinger** [comment]
`User.profileAsBsonDocument` and `User.customDataAsBsonDocument` will be documented under https://jira.mongodb.org/browse/DOCSP-26752

File: examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/AuthenticationTest.kt [hunk 1]
```diff
@@ -186,6 +186,24 @@ class AuthenticationTest: RealmTest() {
 
     }
 
+    @Test
+    fun customFunctionTest() {
+        val app: App = App.create(FLEXIBLE_APP_ID)
```
Comment (dacharyc) [inline examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/AuthenticationTest.kt:191]: Most of the other examples on the Authenticate Users page show the `App.create` and `runBlocking` lines in the code example. For consistency with the other examples, you may want to expand this snippet to include all of ln 191 through 204.

File: examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/AuthenticationTest.kt [hunk 2]
```diff
@@ -186,6 +186,24 @@ class AuthenticationTest: RealmTest() {
...
+            // :snippet-start: custom-function-authentication
+            val customCredentials = Credentials.customFunction(
+                payload = mapOf("username" to "bob")
+            )
+
+            // Pass the generated credential to app.login()
+            val currentUser = app.login(customCredentials)
+            // :snippet-end:
+            // :snippet-start: retrieve-current-user
+            val user = app.currentUser
```
Comment (dacharyc) [inline examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/AuthenticationTest.kt:202]: If you do choose to expand the code example to show the entire `runBlocking` block for consistency with the other examples on this page, we should generate this snippet from somewhere else.

File: source/sdk/kotlin/app-services/authenticate-users.txt [hunk 3]
```diff
@@ -58,15 +81,15 @@ and then pass the generated credential to
 
 .. literalinclude:: /examples/generated/kotlin/AuthenticationTest.snippet.anonymous-authentication.kt
    :language: kotlin
-   :copyable: false
 
 You can optionally specify whether the anonymous user should be reused by 
-passing ``reuseExisting = true`` to ``Credentials.anonymous()`` as long as the anonymous user being reused hasn't logged out.
-The ``reuseExisting`` parameter defaults to false, which creates a new anonymous user credential rather than reusing an existing one.
+passing ``reuseExisting = true`` to ``Credentials.anonymous()`` as long as the anonymous user
```
Comment (dacharyc) [inline source/sdk/kotlin/app-services/authenticate-users.txt:86]: This is a little confusing to me. We're saying here that if a user hasn't logged out, we can reuse their credentials to log the same user in. But why are we logging them in if they haven't logged out?

This implies to me that for some reason, we can't just check for `app.currentUser` and proceed with program execution, but instead have to log in the user again even if they're already logged in. I don't think that's what we mean here. I wonder if we can add some more details around the use case for `reuseExisting` to make it clearer when/why you might use it?

File: source/sdk/kotlin/app-services/authenticate-users.txt [hunk 4]
```diff
@@ -84,25 +107,23 @@ to `app.login()
...
-~~~
+Custom JWT
+~~~~~~~~~~
 
 The :ref:`Custom JWT authentication provider <custom-jwt-authentication>`
 enables users to log in to your application with a custom JSON Web Token.
 To log in with JWT authentication, create a
 JWT credential by calling `Credentials.jwt()
-<{+kotlin-sync-prefix+}io.realm.kotlin.mongodb/-credentials/-companion/facebook.html>`__
+<{+kotlin-sync-prefix+}io.realm.kotlin.mongodb/-credentials/-companion/jwt.html>`__
```
Comment (dacharyc) [inline source/sdk/kotlin/app-services/authenticate-users.txt:120]: Nice catch!

**Review by dacharyc — APPROVED** [review]
Overall, LGTM. But I wonder if we want to tweak the code example for consistency, and I do have a question about the wording of `reuseExisting` for `Credentials.anonymous()`.

File: examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/AuthenticationTest.kt (hunk 1 above)
Comment (cbullinger) [reply examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/AuthenticationTest.kt:191]: I honestly was leaning the other way -- taking out the `App.create` and `runBlocking` lines in the other code examples. I have a ticket to revisit and clean up some of these messier tests; I was planning on updating the snippet scopes then. (I can also confirm with the sdk team to see if they have strong feelings either way.)

File: source/sdk/kotlin/app-services/authenticate-users.txt (hunk 3 above)
Comment (cbullinger) [reply source/sdk/kotlin/app-services/authenticate-users.txt:86]: Excellent question. Getting clarification now and will update with better use case info.

Comment (cbullinger) [reply source/sdk/kotlin/app-services/authenticate-users.txt]: Great catch -- this was incorrectly documented. I've updated the wording and code sample

**Pull request merged**